File: src/weaponsprocurement/core/WeaponStockRecord.java
package weaponsprocurement.core;

import com.fs.starfarer.api.loading.FighterWingSpecAPI;
import com.fs.starfarer.api.loading.MissileSpecAPI;
import com.fs.starfarer.api.loading.ProjectileWeaponSpecAPI;
import com.fs.starfarer.api.loading.WeaponSpecAPI;

import java.util.Collections;
import java.util.List;

public final class WeaponStockRecord {
    private final StockItemType itemType;
    private final String itemId;
    private final String itemKey;
    private final String displayName;
    private final WeaponSpecAPI spec;
    private final FighterWingSpecAPI wingSpec;
    private final int ownedCount;
    private final int playerCargoCount;
    private final int purchasableCount;
    private final int desiredCount;
    private final StockCategory category;
    private final List<SubmarketWeaponStock> submarketStocks;

    public WeaponStockRecord(String weaponId,
                             String displayName,
                             WeaponSpecAPI spec,
                             int ownedCount,
                             int playerCargoCount,
                             int purchasableCount,
                             int desiredCount,
                             StockCategory category,
                             List<SubmarketWeaponStock> submarketStocks) {
        this(StockItemType.WEAPON,
                weaponId,
                displayName,
                spec,
                null,
                ownedCount,
                playerCargoCount,
                purchasableCount,
                desiredCount,
                category,
                submarketStocks);
    }

    public WeaponStockRecord(StockItemType itemType,
                             String itemId,
                             String displayName,
                             WeaponSpecAPI spec,
                             FighterWingSpecAPI wingSpec,
                             int ownedCount,
                             int playerCargoCount,
                             int purchasableCount,
                             int desiredCount,
                             StockCategory category,
                             List<SubmarketWeaponStock> submarketStocks) {
        this.itemType = itemType == null ? StockItemType.WEAPON : itemType;
        this.itemId = itemId;
        this.itemKey = this.itemType.key(itemId);
        this.displayName = displayName;
        this.spec = spec;
        this.wingSpec = wingSpec;
        this.ownedCount = ownedCount;
        this.playerCargoCount = playerCargoCount;
        this.purchasableCount = purchasableCount;
        this.desiredCount = desiredCount;
        this.category = category;
        this.submarketStocks = Collections.unmodifiableList(submarketStocks);
    }

    public String getItemKey() {
        return itemKey;
    }

    public String getItemId() {
        return itemId;
    }

    public StockItemType getItemType() {
        return itemType;
    }

    public boolean isWing() {
        return StockItemType.WING.equals(itemType);
    }

    public String getDisplayName() {
        return displayName;
    }

    public WeaponSpecAPI getSpec() {
        return spec;
    }

    public FighterWingSpecAPI getWingSpec() {
        return wingSpec;
    }

    public int getOwnedCount() {
        return ownedCount;
    }

    public int getStorageCount() {
        return ownedCount;
    }

    public int getPlayerCargoCount() {
        return playerCargoCount;
    }

    public int getPurchasableCount() {
        return purchasableCount;
    }

    public int getBuyableCount() {
        int count = 0;
        for (SubmarketWeaponStock stock : submarketStocks) {
            if (stock.isPurchasable()) {
                count += stock.getCount();
            }
        }
        return count;
    }

    public int getNeededCount() {
        return Math.max(0, desiredCount - ownedCount);
    }

    public int getCheapestPurchasableUnitPrice() {
        int cheapest = Integer.MAX_VALUE;
        for (int i = 0; i < submarketStocks.size(); i++) {
            SubmarketWeaponStock stock = submarketStocks.get(i);
            if (!stock.isPurchasable() || stock.getCount() <= 0) {
                continue;
            }
            cheapest = Math.min(cheapest, stock.getUnitPrice());
        }
        return cheapest == Integer.MAX_VALUE ? Integer.MAX_VALUE : cheapest;
    }

    public int getDesiredCount() {
        return desiredCount;
    }

    public StockCategory getCategory() {
        return category;
    }

    public List<SubmarketWeaponStock> getSubmarketStocks() {
        return submarketStocks;
    }

    public String getSizeLabel() {
        if (isWing()) {
            return "Wing";
        }
        return valueOrUnknown(spec == null ? null : spec.getSize());
    }

    public String getTypeLabel() {
        if (isWing()) {
            return valueOrUnknown(wingSpec == null ? null : wingSpec.getRole());
        }
        return valueOrUnknown(spec == null ? null : spec.getType());
    }

    public String getPrimaryRoleLabel() {
        return valueOrUnknown(spec == null ? null : spec.getPrimaryRoleStr());
    }

    public String getOpCostLabel() {
        return spec == null ? "?" : String.valueOf(Math.round(spec.getOrdnancePointCost(null)));
    }

    public String getDamageTypeLabel() {
        return valueOrUnknown(spec == null ? null : spec.getDamageType());
    }

    public String getDamageLabel() {
        if (isWing()) {
            return "?";
        }
        return spec == null || spec.getDerivedStats() == null ? "?" : String.valueOf(Math.round(spec.getDerivedStats().getDamagePerShot()));
    }

    public String getEmpLabel() {
        if (isWing()) {
            return "?";
        }
        return spec == null || spec.getDerivedStats() == null ? "?" : String.valueOf(Math.round(spec.getDerivedStats().getEmpPerShot()));
    }

    public String getRangeLabel() {
        if (isWing()) {
            return wingSpec == null ? "?" : String.valueOf(Math.round(wingSpec.getRange()));
        }
        return spec == null ? "?" : String.valueOf(Math.round(spec.getMaxRange()));
    }

    public String getFluxPerSecondLabel() {
        if (isWing()) {
            return "?";
        }
        return spec == null || spec.getDerivedStats() == null ? "?" : String.valueOf(Math.round(spec.getDerivedStats().getFluxPerSecond()));
    }

    public String getFluxPerDamageLabel() {
        if (isWing()) {
            return "?";
        }
        return spec == null || spec.getDerivedStats() == null ? "?" : formatOneDecimal(spec.getDerivedStats().getFluxPerDam());
    }

    public String getRefireSecondsLabel() {
        ProjectileWeaponSpecAPI projectile = projectileWeaponSpec();
        return projectile == null ? "?" : formatTwoDecimals(projectile.getRefireDelay());
    }

    public String getSustainedDamagePerSecondLabel() {
        if (spec == null || spec.getDerivedStats() == null) {
            return "?";
        }
        int sustained = Math.round(spec.getDerivedStats().getSustainedDps());
        int burst = Math.round(spec.getDerivedStats().getDps());
        return sustained == burst ? String.valueOf(sustained) : sustained + " (" + burst + ")";
    }

    public boolean hasDifferentSustainedDamagePerSecond() {
        return spec != null
                && spec.getDerivedStats() != null
                && Math.round(spec.getDerivedStats().getSustainedDps()) != Math.round(spec.getDerivedStats().getDps());
    }

    public String getSustainedFluxPerSecondLabel() {
        if (spec == null || spec.getDerivedStats() == null) {
            return "?";
        }
        int sustained = Math.round(spec.getDerivedStats().getSustainedFluxPerSecond());
        int burst = Math.round(spec.getDerivedStats().getFluxPerSecond());
        return sustained == burst ? String.valueOf(sustained) : sustained + " (" + burst + ")";
    }

    public boolean hasDifferentSustainedFluxPerSecond() {
        return spec != null
                && spec.getDerivedStats() != null
                && Math.round(spec.getDerivedStats().getSustainedFluxPerSecond()) != Math.round(spec.getDerivedStats().getFluxPerSecond());
    }

    public String getSustainedEmpPerSecondLabel() {
        return spec == null || spec.getDerivedStats() == null ? "?" : String.valueOf(Math.round(spec.getDerivedStats().getEmpPerSecond()));
    }

    public boolean hasDifferentSustainedEmpPerSecond() {
        return false;
    }

    public String getFluxPerEmpLabel() {
        if (spec == null || spec.getDerivedStats() == null || spec.getDerivedStats().getEmpPerSecond() <= 0f) {
            return "?";
        }
        return String.valueOf(Math.round(spec.getDerivedStats().getSustainedFluxPerSecond() / spec.getDerivedStats().getEmpPerSecond()));
    }

    public String getBeamDpsLabel() {
        return spec == null || !spec.isBeam() || spec.getDerivedStats() == null ? "?" : String.valueOf(Math.round(spec.getDerivedStats().getDps()));
    }

    public String getBeamChargeUpLabel() {
        return spec == null || spec.getBeamChargeupTime() <= 0f ? "?" : formatTwoDecimals(spec.getBeamChargeupTime());
    }

    public String getBeamChargeDownLabel() {
        return spec == null || spec.getBeamChargedownTime() <= 0f ? "?" : formatTwoDecimals(spec.getBeamChargedownTime());
    }

    public String getBurstDelayLabel() {
        ProjectileWeaponSpecAPI projectile = projectileWeaponSpec();
        return projectile == null || projectile.getBurstDelay() <= 0f ? "?" : formatTwoDecimals(projectile.getBurstDelay());
    }

    public String getTurnRateLabel() {
        if (spec == null || spec.getTurnRate() <= 0f) {
            return "?";
        }
        return Math.round(spec.getTurnRate()) + "\u00b0/s";
    }

    public String getMinSpreadLabel() {
        return spec == null || spec.getMinSpread() <= 0f ? "?" : formatOneDecimal(spec.getMinSpread());
    }

    public String getMaxSpreadLabel() {
        return spec == null || spec.getMaxSpread() <= 0f ? "?" : formatOneDecimal(spec.getMaxSpread());
    }

    public String getSpreadPerShotLabel() {
        return spec == null || spec.getSpreadBuildup() <= 0f ? "?" : formatOneDecimal(spec.getSpreadBuildup());
    }

    public String getSpreadDecayLabel() {
        return spec == null || spec.getSpreadDecayRate() <= 0f ? "?" : formatOneDecimal(spec.getSpreadDecayRate());
    }

    public String getProjectileSpeedLabel() {
        ProjectileWeaponSpecAPI projectile = projectileWeaponSpec();
        if (projectile == null) {
            return "?";
        }
        float speed;
        try {
            speed = projectile.getProjectileSpeed(null, null);
        } catch (RuntimeException ex) {
            return "?";
        }
        return speed <= 0f || Float.isNaN(speed) || Float.isInfinite(speed)
                ? "?"
                : String.valueOf(Math.round(speed));
    }

    public String getLaunchSpeedLabel() {
        MissileSpecAPI missile = missileSpec();
        return missile == null || missile.getLaunchSpeed() <= 0f ? "?" : String.valueOf(Math.round(missile.getLaunchSpeed()));
    }

    public String getFlightTimeLabel() {
        MissileSpecAPI missile = missileSpec();
        return missile == null || missile.getMaxFlightTime() <= 0f ? "?" : formatTwoDecimals(missile.getMaxFlightTime());
    }

    public String getGuidedLabel() {
        MissileSpecAPI missile = missileSpec();
        if (missile == null) {
            return "?";
        }
        String tracking = spec == null ? null : spec.getTrackingStr();
        boolean guided = tracking != null && tracking.trim().length() > 0 && !"None".equalsIgnoreCase(tracking.trim());
        return guided ? "TRUE" : "FALSE";
    }

    public String getMaxAmmoLabel() {
        return spec == null || !spec.usesAmmo() || spec.getMaxAmmo() <= 0 ? "?" : String.valueOf(spec.getMaxAmmo());
    }

    public String getSecPerReloadLabel() {
        if (spec == null || !spec.usesAmmo() || spec.getAmmoPerSecond() <= 0f || spec.getReloadSize() <= 0f) {
            return "?";
        }
        return formatTwoDecimals(spec.getReloadSize() / spec.getAmmoPerSecond());
    }

    public String getAmmoGainLabel() {
        return spec == null || !spec.usesAmmo() || spec.getAmmoPerSecond() <= 0f ? "?" : formatOneDecimal(spec.getAmmoPerSecond());
    }

    public String getAccuracyLabel() {
        return valueOrUnknown(spec == null ? null : spec.getAccuracyStr());
    }

    public String getWingFighterCountLabel() {
        return wingSpec == null ? "?" : String.valueOf(wingSpec.getNumFighters());
    }

    public String getWingOpCostLabel() {
        return wingSpec == null ? "?" : String.valueOf(Math.round(wingSpec.getOpCost(null)));
    }

    public String getWingRefitTimeLabel() {
        return wingSpec == null ? "?" : formatOneDecimal(wingSpec.getRefitTime()) + "s";
    }

    private static String formatOneDecimal(float value) {
        if (Float.isNaN(value) || Float.isInfinite(value)) {
            return "?";
        }
        return String.valueOf(Math.round(value * 10f) / 10f);
    }

    private static String formatTwoDecimals(float value) {
        if (Float.isNaN(value) || Float.isInfinite(value)) {
            return "?";
        }
        return String.format(java.util.Locale.US, "%.2f", value);
    }

    private ProjectileWeaponSpecAPI projectileWeaponSpec() {
        return spec instanceof ProjectileWeaponSpecAPI ? (ProjectileWeaponSpecAPI) spec : null;
    }

    private MissileSpecAPI missileSpec() {
        Object projectile = spec == null ? null : spec.getProjectileSpec();
        return projectile instanceof MissileSpecAPI ? (MissileSpecAPI) projectile : null;
    }

    private static String valueOrUnknown(Object value) {
        return value == null ? "?" : String.valueOf(value);
    }
}

File: src/main/kotlin/weaponsprocurement/core/WeaponStockRecord.kt
package weaponsprocurement.core

import com.fs.starfarer.api.loading.FighterWingSpecAPI
import com.fs.starfarer.api.loading.MissileSpecAPI
import com.fs.starfarer.api.loading.ProjectileWeaponSpecAPI
import com.fs.starfarer.api.loading.WeaponSpecAPI
import java.util.Collections
import java.util.Locale

class WeaponStockRecord(
    itemType: StockItemType?,
    val itemId: String?,
    val displayName: String?,
    val spec: WeaponSpecAPI?,
    val wingSpec: FighterWingSpecAPI?,
    val ownedCount: Int,
    val playerCargoCount: Int,
    val purchasableCount: Int,
    val desiredCount: Int,
    val category: StockCategory?,
    submarketStocks: List<SubmarketWeaponStock>,
    private val fixerRarity: FixerRarity?,
) {
    constructor(
        weaponId: String?,
        displayName: String?,
        spec: WeaponSpecAPI?,
        ownedCount: Int,
        playerCargoCount: Int,
        purchasableCount: Int,
        desiredCount: Int,
        category: StockCategory?,
        submarketStocks: List<SubmarketWeaponStock>,
    ) : this(
        StockItemType.WEAPON,
        weaponId,
        displayName,
        spec,
        null,
        ownedCount,
        playerCargoCount,
        purchasableCount,
        desiredCount,
        category,
        submarketStocks,
        null,
    )

    constructor(
        itemType: StockItemType?,
        itemId: String?,
        displayName: String?,
        spec: WeaponSpecAPI?,
        wingSpec: FighterWingSpecAPI?,
        ownedCount: Int,
        playerCargoCount: Int,
        purchasableCount: Int,
        desiredCount: Int,
        category: StockCategory?,
        submarketStocks: List<SubmarketWeaponStock>,
    ) : this(
        itemType,
        itemId,
        displayName,
        spec,
        wingSpec,
        ownedCount,
        playerCargoCount,
        purchasableCount,
        desiredCount,
        category,
        submarketStocks,
        null,
    )

    val itemType: StockItemType = itemType ?: StockItemType.WEAPON
    val itemKey: String = this.itemType.key(itemId)
    val submarketStocks: List<SubmarketWeaponStock> = Collections.unmodifiableList(submarketStocks)

    fun isWing(): Boolean = StockItemType.WING == itemType

    val storageCount: Int
        get() = ownedCount

    val buyableCount: Int
        get() {
            var count = 0
            for (stock in submarketStocks) {
                if (stock.isPurchasable()) count += stock.count
            }
            return count
        }

    val neededCount: Int
        get() = Math.max(0, desiredCount - ownedCount)

    val cheapestPurchasableUnitPrice: Int
        get() {
            var cheapest = Int.MAX_VALUE
            for (stock in submarketStocks) {
                if (!stock.isPurchasable() || stock.count <= 0) continue
                cheapest = Math.min(cheapest, stock.unitPrice)
            }
            return cheapest
        }

    val fixerRarityLabel: String?
        get() = fixerRarity?.label

    val sizeLabel: String
        get() {
            if (isWing()) return "Wing"
            return valueOrUnknown(spec?.size)
        }

    val typeLabel: String
        get() {
            if (isWing()) return valueOrUnknown(wingSpec?.role)
            return valueOrUnknown(spec?.type)
        }

    val primaryRoleLabel: String
        get() = valueOrUnknown(spec?.primaryRoleStr)

    val opCostLabel: String
        get() = if (spec == null) "?" else Math.round(spec.getOrdnancePointCost(null)).toString()

    val damageTypeLabel: String
        get() = valueOrUnknown(spec?.damageType)

    val damageLabel: String
        get() {
            if (isWing()) return "?"
            return if (spec?.derivedStats == null) "?" else Math.round(spec.derivedStats.damagePerShot).toString()
        }

    val empLabel: String
        get() {
            if (isWing()) return "?"
            return if (spec?.derivedStats == null) "?" else Math.round(spec.derivedStats.empPerShot).toString()
        }

    val rangeLabel: String
        get() {
            if (isWing()) return if (wingSpec == null) "?" else Math.round(wingSpec.range).toString()
            return if (spec == null) "?" else Math.round(spec.maxRange).toString()
        }

    val fluxPerSecondLabel: String
        get() = if (isWing() || spec?.derivedStats == null) {
            "?"
        } else {
            Math.round(spec.derivedStats.fluxPerSecond).toString()
        }

    val fluxPerDamageLabel: String
        get() = if (isWing() || spec?.derivedStats == null) {
            "?"
        } else {
            formatOneDecimal(spec.derivedStats.fluxPerDam)
        }

    val refireSecondsLabel: String
        get() {
            val projectile = projectileWeaponSpec()
            return if (projectile == null) "?" else formatTwoDecimals(projectile.refireDelay)
        }

    val sustainedDamagePerSecondLabel: String
        get() {
            if (spec?.derivedStats == null) return "?"
            val sustained = Math.round(spec.derivedStats.sustainedDps)
            val burst = Math.round(spec.derivedStats.dps)
            return if (sustained == burst) sustained.toString() else "$sustained ($burst)"
        }

    fun hasDifferentSustainedDamagePerSecond(): Boolean {
        return spec?.derivedStats != null &&
            Math.round(spec.derivedStats.sustainedDps) != Math.round(spec.derivedStats.dps)
    }

    val sustainedFluxPerSecondLabel: String
        get() {
            if (spec?.derivedStats == null) return "?"
            val sustained = Math.round(spec.derivedStats.sustainedFluxPerSecond)
            val burst = Math.round(spec.derivedStats.fluxPerSecond)
            return if (sustained == burst) sustained.toString() else "$sustained ($burst)"
        }

    fun hasDifferentSustainedFluxPerSecond(): Boolean {
        return spec?.derivedStats != null &&
            Math.round(spec.derivedStats.sustainedFluxPerSecond) != Math.round(spec.derivedStats.fluxPerSecond)
    }

    val sustainedEmpPerSecondLabel: String
        get() = if (spec?.derivedStats == null) "?" else Math.round(spec.derivedStats.empPerSecond).toString()

    fun hasDifferentSustainedEmpPerSecond(): Boolean = false

    val fluxPerEmpLabel: String
        get() {
            val stats = spec?.derivedStats ?: return "?"
            if (stats.empPerSecond <= 0f) return "?"
            return Math.round(stats.sustainedFluxPerSecond / stats.empPerSecond).toString()
        }

    val beamDpsLabel: String
        get() = if (spec == null || !spec.isBeam || spec.derivedStats == null) {
            "?"
        } else {
            Math.round(spec.derivedStats.dps).toString()
        }

    val beamChargeUpLabel: String
        get() = if (spec == null || spec.beamChargeupTime <= 0f) "?" else formatTwoDecimals(spec.beamChargeupTime)

    val beamChargeDownLabel: String
        get() = if (spec == null || spec.beamChargedownTime <= 0f) "?" else formatTwoDecimals(spec.beamChargedownTime)

    val burstDelayLabel: String
        get() {
            val projectile = projectileWeaponSpec()
            return if (projectile == null || projectile.burstDelay <= 0f) "?" else formatTwoDecimals(projectile.burstDelay)
        }

    val turnRateLabel: String
        get() {
            if (spec == null || spec.turnRate <= 0f) return "?"
            return Math.round(spec.turnRate).toString() + "\u00b0/s"
        }

    val minSpreadLabel: String
        get() = if (spec == null || spec.minSpread <= 0f) "?" else formatOneDecimal(spec.minSpread)

    val maxSpreadLabel: String
        get() = if (spec == null || spec.maxSpread <= 0f) "?" else formatOneDecimal(spec.maxSpread)

    val spreadPerShotLabel: String
        get() = if (spec == null || spec.spreadBuildup <= 0f) "?" else formatOneDecimal(spec.spreadBuildup)

    val spreadDecayLabel: String
        get() = if (spec == null || spec.spreadDecayRate <= 0f) "?" else formatOneDecimal(spec.spreadDecayRate)

    val projectileSpeedLabel: String
        get() {
            val projectile = projectileWeaponSpec() ?: return "?"
            val speed = try {
                projectile.getProjectileSpeed(null, null)
            } catch (_: RuntimeException) {
                return "?"
            }
            return if (speed <= 0f || speed.isNaN() || speed.isInfinite()) "?" else Math.round(speed).toString()
        }

    val launchSpeedLabel: String
        get() {
            val missile = missileSpec()
            return if (missile == null || missile.launchSpeed <= 0f) "?" else Math.round(missile.launchSpeed).toString()
        }

    val flightTimeLabel: String
        get() {
            val missile = missileSpec()
            return if (missile == null || missile.maxFlightTime <= 0f) "?" else formatTwoDecimals(missile.maxFlightTime)
        }

    val guidedLabel: String
        get() {
            if (missileSpec() == null) return "?"
            val tracking = spec?.trackingStr
            val guided = !tracking.isNullOrBlank() && !"None".equals(tracking.trim(), ignoreCase = true)
            return if (guided) "TRUE" else "FALSE"
        }

    val maxAmmoLabel: String
        get() = if (spec == null || !spec.usesAmmo() || spec.maxAmmo <= 0) "?" else spec.maxAmmo.toString()

    val secPerReloadLabel: String
        get() {
            if (spec == null || !spec.usesAmmo() || spec.ammoPerSecond <= 0f || spec.reloadSize <= 0f) return "?"
            return formatTwoDecimals(spec.reloadSize / spec.ammoPerSecond)
        }

    val ammoGainLabel: String
        get() = if (spec == null || !spec.usesAmmo() || spec.ammoPerSecond <= 0f) {
            "?"
        } else {
            formatOneDecimal(spec.ammoPerSecond)
        }

    val accuracyLabel: String
        get() = valueOrUnknown(spec?.accuracyStr)

    val wingFighterCountLabel: String
        get() = wingSpec?.numFighters?.toString() ?: "?"

    val wingOpCostLabel: String
        get() = if (wingSpec == null) "?" else Math.round(wingSpec.getOpCost(null)).toString()

    val wingRefitTimeLabel: String
        get() = if (wingSpec == null) "?" else formatOneDecimal(wingSpec.refitTime) + "s"

    private fun projectileWeaponSpec(): ProjectileWeaponSpecAPI? {
        return if (spec is ProjectileWeaponSpecAPI) spec else null
    }

    private fun missileSpec(): MissileSpecAPI? {
        val projectile = spec?.projectileSpec
        return if (projectile is MissileSpecAPI) projectile else null
    }

    companion object {
        private fun formatOneDecimal(value: Float): String {
            if (value.isNaN() || value.isInfinite()) return "?"
            return (Math.round(value * 10f) / 10f).toString()
        }

        private fun formatTwoDecimals(value: Float): String {
            if (value.isNaN() || value.isInfinite()) return "?"
            return String.format(Locale.US, "%.2f", value)
        }

        private fun valueOrUnknown(value: Any?): String {
            return value?.toString() ?: "?"
        }
    }
}
